Fixes for sliceviewer line plots

martyngigg opened the pull request:
**Description of work.**

Several fixes to address problems with the line plots in sliceviewer:
* avoiding an error with setting normalization options too early
* disables autoscaling of lineplots on axis linked to image. This avoids the image plot immediately zooming out when the mouse moves over the image.
* fixes the home button to simply reset the data limits rather than use the matplotlib navstack

**To test:**

# Simple test of lines
```python
# import mantid algorithms, numpy and matplotlib
from mantid.simpleapi import *
import matplotlib.pyplot as plt
import numpy as np
nrows, ncols = 5, 5
signal = np.arange(0,nrows*ncols)
x = np.arange(0,2*ncols+1,step=2)
ws = CreateWorkspace(DataX=x, DataY=signal, NSpec=nrows,UnitX='TOF')
```
creates a simple workspace with increasing values. Check this plots okay, the values displayed in the cursor coordinates are correct and the line plots make sense as the mouse moves around.

# Test with MD workspace

The files are in the [training data](http://sourceforge.net/projects/mantid/files/Sample%20Data/TrainingCourseData.zip/download)
```
SXD23767 = Load(Filename='SXD23767.raw', LoadMonitors='Exclude')
mdws = ConvertToMD(
    InputWorkspace=SXD23767,
    QDimensions="Q3D",
    dEAnalysisMode="Elastic",
    QConversionScales="Q in A^-1",
    LorentzCorrection='1',
    MinValues=[-15, -15, -15],
    MaxValues=[15, 15, 15],
    SplitInto='2',
    SplitThreshold='50',
    MaxRecursionDepth='14')
peaksws = Load('peaks_qLab.nxs')
```
* show the sliceviewer on `mdws`
* click the peaks overlay and select `peaksws`
* in the table click on the row `123` that should zoom to a peak.
* activate the line plots and move the mouse around over the squares. The peaks in the lines should be in the center of the pixel.

Refs #28579.

*This does not require release notes* because **it fixes a regression**

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

martyngigg commented:
Some conflicts have developed with the non-orthogonal axes PR that was just merged. I'm going to rebase and fix them now.

gemmaguest commented:
Should it work with panning? If I enable line plots, zoom in, then pan, I get this:

![image](https://user-images.githubusercontent.com/12895056/81945065-80e15a80-95f5-11ea-8c74-d59917ca478c.png)

martyngigg commented:
Thanks for testing this.
I know that looks odd but is that the edge of the data? It's panned the line plots to keep in line with the data, which is what we want I think. I've tried with a larger workspace and if you zoom in around the middle and pan it drags the line plots with it so the peaks are in the correct place.

gemmaguest commented:
> 
> 
> Thanks for testing this.
> I know that looks odd but is that the edge of the data? It's panned the line plots to keep in line with the data, which is what we want I think. I've tried with a larger workspace and if you zoom in around the middle and pan it drags the line plots with it so the peaks are in the correct place.

Ah, yes, that's right. I think I'd prefer if it didn't let me pan outside the data but it's not a big issue. Also once I've panned outside the original area, the Home button resets to include the extra area I panned to rather than the original area. I doubt that's related to this PR though. The line plots still always show correctly, anyway :+1:

gemmaguest commented:
May not be related to this PR but if I attempt to overlap the `peaksws` onto the non-MD `ws` from the first example I get an unhandled exception.

martyngigg commented:
> > Thanks for testing this.
> > I know that looks odd but is that the edge of the data? It's panned the line plots to keep in line with the data, which is what we want I think. I've tried with a larger workspace and if you zoom in around the middle and pan it drags the line plots with it so the peaks are in the correct place.
> 
> Ah, yes, that's right. I think I'd prefer if it didn't let me pan outside the data but it's not a big issue. Also once I've panned outside the original area, the Home button resets to include the extra area I panned to rather than the original area. I doubt that's related to this PR though. The line plots still always show correctly, anyway +1

Agreed it would be better if it stopped scrolling but it may be more trouble than it is worth to try and stop matplotlib from doing this.


> Not related to this PR but just documenting things I find:
> 
>     * If I attempt to overlap the `peaksws` onto the non-MD `ws` from the first example I get an unhandled exception.
> 
>     * The grid toggle button doesn't stay depressed when on.
> 
>     * Turning line plots on turns off the grid (not sure if that's intentional).
> 
>     * The cursor flickers between different shapes when it's calculating the line plots. Not a big issue but it's a bit distracting.
> 
>     * In the second example, if I click Home then click on the already selected row in the table, nothing happens, but when I mouse over the plot again, it zooms to the peak. Clicking any (non-selected) row immediately zooms to the peak.
> 
>     * If I toggle the axes when lineplots are enabled, the axes don't update till I move the mouse over the plot again, and the lineplots look odd:

Thanks for picking these things up. I think I will fix the lineplot related ones here and move the others to separate issue(s).

martyngigg commented:
I've fixed the following here:
* Turning line plots on turns off the grid (not sure if that's intentional).
* If I toggle the axes when lineplots are enabled, the axes don't update till I move the mouse over the plot again, and the lineplots look odd

> If I attempt to overlap the peaksws onto the non-MD ws from the first example I get an unhandled exception.

I'll deal with this in a separate issue too.

> The grid toggle button doesn't stay depressed when on.

I attempted to fix this in another issue and it got surprisingly complicated as matplotlib only allows toggling grids and the state management with the button and the nonorthogonal axes (where we always want it on) was quite confusing. I'll open a separate lower priority issue for this.

> The cursor flickers between different shapes when it's calculating the line plots. Not a big issue but it's a bit distracting.

I agree this is annoying but this is buried in the matplotlib rendering and appears to have been fixed in [version 3.2.0](https://github.com/matplotlib/matplotlib/pull/9672/commits/0944dcbe6dd755fd51a91f9ca2eb57c4ae798624). We can at least update to this on Windows/macOS.

gemmaguest commented:
Getting an unhandled exception. A bit hard to reproduce but I think this does it:
- Enable line plots
- Zoom in to top right corner
- Pan up and right to show white space (i.e. outside the data)
- Toggle the axes - it takes quite a long time and then the plot goes blank
- Mouse over the blank plot (nothing seems to happen but I think this is necessary to reproduce the bug)
- Toggle the axes again


```
Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/matplotlib/cbook/__init__.py", line 388, in process
    proxy(*args, **kwargs)
  File "/usr/lib/python3/dist-packages/matplotlib/cbook/__init__.py", line 228, in __call__
    return mtd(*args, **kwargs)
  File "/home/gemma/work/checkout/mantid2/qt/python/mantidqt/widgets/sliceviewer/view.py", line 356, in mouse_outside_image
    self.delete_line_plot_lines()
  File "/home/gemma/work/checkout/mantid2/qt/python/mantidqt/widgets/sliceviewer/view.py", line 307, in delete_line_plot_lines
    self.xfig.remove()
  File "/usr/lib/python3/dist-packages/matplotlib/artist.py", line 151, in remove
    self._remove_method(self)
  File "/usr/lib/python3/dist-packages/matplotlib/axes/_base.py", line 1783, in <lambda>
    line._remove_method = lambda h: self.lines.remove(h)
ValueError: list.remove(x): x not in list
```

gemmaguest commented:
The line plots for mdws don't look right if I click on the peak in the table. If I zoom out and back in they are better but the horizontal plot is still not centred. In both of these I had the mouse over the yellow pixel.

![image](https://user-images.githubusercontent.com/12895056/82208718-ee49ff80-9903-11ea-9b93-25e85cf83fbe.png)

![image](https://user-images.githubusercontent.com/12895056/82208666-d6727b80-9903-11ea-92c3-fb352ba7cedb.png)

martyngigg commented:
@gemmaguest Thanks for giving this a bash around. I've added protection for that possible `ValueError` you found.
The issue with the misleading line plots is a little more subtle. As mentioned on Slack, when zoomed out the line plot should be fine. The issue comes on zooming.
This likely to be changed with #28689 that enables dynamic rebinning. Would it be acceptable to leave addressing this to that issue?

gemmaguest commented:
It no longer throws the `ValueError` but following the same steps I now get a weird state where the TOF axis has turned into `Spectrum` and the line plot doesn't show:

![image](https://user-images.githubusercontent.com/12895056/82349076-92f63b00-99f1-11ea-84ab-d4f1b2864317.png)

martyngigg commented:
I think this was caused by a bad interaction with my other PR. I've rebased after some updates to that and I think this looks better now...

gemmaguest commented:
> 
> 
> I think this was caused by a bad interaction with my other PR. I've rebased after some updates to that and I think this looks better now...

Hmm I'm still seeing the same behaviour. Are you able to reproduce it?

martyngigg commented:
> > I think this was caused by a bad interaction with my other PR. I've rebased after some updates to that and I think this looks better now...
> 
> Hmm I'm still seeing the same behaviour. Are you able to reproduce it?

I thought I was but I just realised when you said "toggle axes" I was switching the lineplot axes on/off :facepalm: I assumed the ValueError was benign as the plot had just removed the line already. I can see the issue properly now. Thanks for being persistent here.

martyngigg commented:
These latest changes should make the pan/zoom behave more sensibly when flipping dimensions for a MatrixWorkspace. There is a performance question around this behaviour with MatrixWorkspace as it continually has to resample as you move/zoom to avoid having to calculate a potentially large image array for large workspaces. For standard colorfill plots the interactivity is better but it needs to compute the full array and for some large datasets this can take minutes.

Something to address in the future in #28693

gemmaguest commented:
Flipping dimensions works much better now, but I got an unhandled exception once when playing around with this and zooming. I haven't managed to reproduce it yet but here's the error:

```
Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/matplotlib/cbook/__init__.py", line 388, in process
    proxy(*args, **kwargs)
  File "/usr/lib/python3/dist-packages/matplotlib/cbook/__init__.py", line 228, in __call__
    return mtd(*args, **kwargs)
  File "/home/gemma/work/checkout/mantid2/qt/python/mantidqt/widgets/sliceviewer/view.py", line 359, in mouse_outside_image
    self.delete_line_plot_lines()
  File "/home/gemma/work/checkout/mantid2/qt/python/mantidqt/widgets/sliceviewer/view.py", line 319, in delete_line_plot_lines
    self.xfig.remove()
  File "/usr/lib/python3/dist-packages/matplotlib/artist.py", line 151, in remove
    self._remove_method(self)
  File "/usr/lib/python3/dist-packages/matplotlib/axes/_base.py", line 1783, in <lambda>
    line._remove_method = lambda h: self.lines.remove(h)
ValueError: list.remove(x): x not in list
Python-[Error] Traceback (most recent call last):
  File "/home/gemma/work/checkout/mantid2/qt/python/mantidqt/widgets/sliceviewer/presenter.py", line 113, in dimensions_changed
    self.new_plot()
  File "/home/gemma/work/checkout/mantid2/qt/python/mantidqt/widgets/sliceviewer/presenter.py", line 73, in new_plot_matrix
    self.view.data_view.plot_matrix(self.model.get_ws(), normalize=self.normalization)
  File "/home/gemma/work/checkout/mantid2/qt/python/mantidqt/widgets/sliceviewer/view.py", line 231, in plot_matrix
    self.draw_plot()
  File "/home/gemma/work/checkout/mantid2/qt/python/mantidqt/widgets/sliceviewer/view.py", line 250, in draw_plot
    self.delete_line_plot_lines()
  File "/home/gemma/work/checkout/mantid2/qt/python/mantidqt/widgets/sliceviewer/view.py", line 319, in delete_line_plot_lines
    self.xfig.remove()
  File "/usr/lib/python3/dist-packages/matplotlib/artist.py", line 151, in remove
    self._remove_method(self)
  File "/usr/lib/python3/dist-packages/matplotlib/axes/_base.py", line 1783, in <lambda>
    line._remove_method = lambda h: self.lines.remove(h)
ValueError: list.remove(x): x not in list
```

Also, flipping dimensions resets the zoom level. Is that expected?

martyngigg commented:
> Flipping dimensions works much better now, but I got an unhandled exception once when playing around with this and zooming. I haven't managed to reproduce it yet but here's the error:
> 
> ```
> Traceback (most recent call last):
>   File "/usr/lib/python3/dist-packages/matplotlib/cbook/__init__.py", line 388, in process
>     proxy(*args, **kwargs)
>   File "/usr/lib/python3/dist-packages/matplotlib/cbook/__init__.py", line 228, in __call__
>     return mtd(*args, **kwargs)
>   File "/home/gemma/work/checkout/mantid2/qt/python/mantidqt/widgets/sliceviewer/view.py", line 359, in mouse_outside_image
>     self.delete_line_plot_lines()
>   File "/home/gemma/work/checkout/mantid2/qt/python/mantidqt/widgets/sliceviewer/view.py", line 319, in delete_line_plot_lines
>     self.xfig.remove()
>   File "/usr/lib/python3/dist-packages/matplotlib/artist.py", line 151, in remove
>     self._remove_method(self)
>   File "/usr/lib/python3/dist-packages/matplotlib/axes/_base.py", line 1783, in <lambda>
>     line._remove_method = lambda h: self.lines.remove(h)
> ValueError: list.remove(x): x not in list
> Python-[Error] Traceback (most recent call last):
>   File "/home/gemma/work/checkout/mantid2/qt/python/mantidqt/widgets/sliceviewer/presenter.py", line 113, in dimensions_changed
>     self.new_plot()
>   File "/home/gemma/work/checkout/mantid2/qt/python/mantidqt/widgets/sliceviewer/presenter.py", line 73, in new_plot_matrix
>     self.view.data_view.plot_matrix(self.model.get_ws(), normalize=self.normalization)
>   File "/home/gemma/work/checkout/mantid2/qt/python/mantidqt/widgets/sliceviewer/view.py", line 231, in plot_matrix
>     self.draw_plot()
>   File "/home/gemma/work/checkout/mantid2/qt/python/mantidqt/widgets/sliceviewer/view.py", line 250, in draw_plot
>     self.delete_line_plot_lines()
>   File "/home/gemma/work/checkout/mantid2/qt/python/mantidqt/widgets/sliceviewer/view.py", line 319, in delete_line_plot_lines
>     self.xfig.remove()
>   File "/usr/lib/python3/dist-packages/matplotlib/artist.py", line 151, in remove
>     self._remove_method(self)
>   File "/usr/lib/python3/dist-packages/matplotlib/axes/_base.py", line 1783, in <lambda>
>     line._remove_method = lambda h: self.lines.remove(h)
> ValueError: list.remove(x): x not in list
> ```

Thanks. I couldn't reproduce this but as protection against something else removing the lines I've included an exception handler for the ValueError.

> Also, flipping dimensions resets the zoom level. Is that expected?

I originally couldn't get this to play well with swapping the axes limits with the resampling that goes on for matrix workspaces. I think I have fixed it now so that when you swap dimensions it also keeps and swaps the data limits that were present before.

gemmaguest commented on qt/python/mantidqt/widgets/sliceviewer/view.py:
```diff
@@ -219,28 +232,45 @@ def plot_matrix(self, ws, **kwargs):
+    def reset_image_view_to_data_limits(self):
+        """Reset the image view limits to encompass all of the data"""
+        e1, e2, e3, e4 = self.image.get_extent()
+        # if self.dimensions.transpose:
```
Should this commented-out code be here?

martyngigg replied:
Thanks for catching this. I meant to remove that after testing something.

gemmaguest commented:
The colour scale changes when I flip the dimensions when zoomed in, which I find a bit disconcerting, although not sure if it's a big issue. This didn't work at all before so probably not worth holding up the PR for.

![sliceviewer_before](https://user-images.githubusercontent.com/12895056/83521055-9223d580-a4d6-11ea-8477-836ce3c5191d.PNG)

![sliceviewer_after](https://user-images.githubusercontent.com/12895056/83521070-97812000-a4d6-11ea-8057-1f0c98d81765.PNG)

gemmaguest commented:
Not related to this PR but if I zoom in far enough the exponential label overlaps the lineplot/axes labels

![sliceviewer_labels](https://user-images.githubusercontent.com/12895056/83522104-14f96000-a4d8-11ea-879b-9ae4d5241aa3.PNG)

gemmaguest commented:
The Home button doesn't work for the MD workspace now, and I get this error when selecting a row in the table:
```
Python-[Error] Traceback (most recent call last):
  File "/home/gemma/work/checkout/mantid2/qt/python/mantidqt/widgets/sliceviewer/presenter.py", line 119, in slicepoint_changed
    PeaksViewerPresenter.Event.SlicePointChanged)
  File "/home/gemma/work/checkout/mantid2/qt/python/mantidqt/widgets/sliceviewer/presenter.py", line 215, in _call_peaks_presenter_if_created
    getattr(self._peaks_presenter)(*args, **kwargs)
TypeError: getattr expected at least 2 arguments, got 1
```

martyngigg commented:
Thanks for persisting with test this. I hope I've addressed the concerns now.

> The colour scale changes when I flip the dimensions when zoomed in, which I find a bit disconcerting, although not sure if it's a big issue. This didn't work at all before so probably not worth holding up the PR for.
> 
> ![sliceviewer_before](https://user-images.githubusercontent.com/12895056/83521055-9223d580-a4d6-11ea-8477-836ce3c5191d.PNG)
> 
> ![sliceviewer_after](https://user-images.githubusercontent.com/12895056/83521070-97812000-a4d6-11ea-8057-1f0c98d81765.PNG)

This one is a little tricky due to the way things are resampled and how the colorbar in matplotlib works. The dynamic rebinning for MD will affect this so I'll see if I can do something about it. 

> Not related to this PR but if I zoom in far enough the exponential label overlaps the lineplot/axes labels
> 
> ![sliceviewer_labels](https://user-images.githubusercontent.com/12895056/83522104-14f96000-a4d8-11ea-879b-9ae4d5241aa3.PNG)

This is a matplotlib "feature" in that it just allows you to zoom indefinitely. We could consider trying to step in but I'd be tempted to see if users really see a problem with normal usage.

> The Home button doesn't work for the MD workspace now, and I get this error when selecting a row in the table:
> 
> ```
> Python-[Error] Traceback (most recent call last):
>   File "/home/gemma/work/checkout/mantid2/qt/python/mantidqt/widgets/sliceviewer/presenter.py", line 119, in slicepoint_changed
>     PeaksViewerPresenter.Event.SlicePointChanged)
>   File "/home/gemma/work/checkout/mantid2/qt/python/mantidqt/widgets/sliceviewer/presenter.py", line 215, in _call_peaks_presenter_if_created
>     getattr(self._peaks_presenter)(*args, **kwargs)
> TypeError: getattr expected at least 2 arguments, got 1
> ```

These should now be fixed.
